File: client/chat_client.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net"
	"os"
)

func main() {
	host := "localhost"
	port := "6666"

	if len(os.Args) > 2 {
		host = os.Args[1]
		port = os.Args[2]
	}

	connect(host, port)
}

func connect(host, port string) {
	address := net.JoinHostPort(host, port)

	conn, err := net.Dial("tcp", address)
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	fmt.Printf("Connected \n")

	go inputLoop(conn)
	messageLoop(conn)
}

func inputLoop(conn net.Conn) {
	stdinScanner := bufio.NewScanner(os.Stdin)
	
	for {
		fmt.Print("> ")
		
		if !stdinScanner.Scan() {
			fmt.Println("DC")
			conn.Close()
			return
		}

		fmt.Fprintf(conn, "%s\n", stdinScanner.Text())
	}
}

func messageLoop(conn net.Conn) {
	connScanner := bufio.NewScanner(conn)

	for {
		// message, err := reader.ReadString('\n')
		if !connScanner.Scan() {
		  break
		}
		clearLine()
		// fmt.Print(message)
		fmt.Printf("%s\n", connScanner.Text())
		fmt.Print("> ")
	}
}

func clearLine() {
	fmt.Print("\r\033[K")
}

File: client/chat_client.exs
defmodule ChatClient do
  def connect(host \\ "localhost", port \\ 6666) do
    {:ok, socket} = :gen_tcp.connect(host, port,
    [:binary, active: true, packet: :line])
    IO.puts("connected to #{host}:#{port}")
    start_input_loop(socket)
    message_loop(socket)
  end

  defp start_input_loop(socket) do
    spawn(fn -> input_loop(socket) end)
  end

  defp input_loop(socket) do
    case IO.gets("> ") do
      :eof ->
        IO.puts("DC")
        :gen_tcp.close(socket)

      data ->
        :gen_tcp.send(socket, data)
        input_loop(socket)
    end
  end

  defp message_loop(socket) do
    receive do
      {:tcp, ^socket, data} ->
        clear_line()
        IO.write(data)

        IO.write("> ")
        message_loop(socket)

      {:tcp_closed, ^socket} ->
        IO.puts("Server closed the connection.")
        :gen_tcp.close(socket)

      {:tcp_error, ^socket, reason} ->
        IO.puts("Error with socket: #{inspect(reason)}")
        :gen_tcp.close(socket)

      other ->
        IO.puts("Unexpected message: #{inspect(other)}")
        message_loop(socket)
    end
  end

  defp clear_line() do
    IO.write("\r\e[K")
  end

end

with [host, port | _] <- System.argv(),
  {port, ""} <- Integer.parse(port)
do
  ChatClient.connect(String.to_atom(host), port)
else
  _ -> ChatClient.connect(:localhost, 6666)
end
